components/script/dom/bindings/refcounted.rs

```diff
@@ -153,59 +132,55 @@ impl LiveDOMReferences {
         });
     }
 
-    fn addref<T: Reflectable>(&self, ptr: *const T) -> Arc<Mutex<usize>> {
+    fn addref<T: Reflectable>(&self, ptr: *const T) -> Arc<TrustedReference> {
         let mut table = self.table.borrow_mut();
+        let capacity = table.capacity();
+        let len = table.len();
+        if (0 < capacity) && (capacity <= len) {
+            info!("growing refcounted references by {}", len);
+            remove_nulls(&mut table);
+            table.reserve(len);
+        }
         match table.entry(ptr as *const libc::c_void) {
-            Occupied(mut entry) => {
-                let refcount = entry.get_mut();
-                *refcount.lock().unwrap() += 1;
-                refcount.clone()
-            }
+            Occupied(mut entry) => match entry.get().upgrade() {
+                Some(refcount) => refcount.clone(),
+                None => {
+                    let refcount = Arc::new(TrustedReference::new(ptr));
+                    entry.insert(Arc::downgrade(&refcount));
+                    refcount
+                },
+            },
             Vacant(entry) => {
-                let refcount = Arc::new(Mutex::new(1));
-                entry.insert(refcount.clone());
+                let refcount = Arc::new(TrustedReference::new(ptr));
+                entry.insert(Arc::downgrade(&refcount));
                 refcount
             }
         }
     }
+}
 
-    /// Unpin the given DOM object if its refcount is 0.
-    pub fn cleanup(raw_reflectable: TrustedReference) {
-        let TrustedReference(raw_reflectable) = raw_reflectable;
-        LIVE_REFERENCES.with(|ref r| {
-            let r = r.borrow();
-            let live_references = r.as_ref().unwrap();
-            let mut table = live_references.table.borrow_mut();
-            match table.entry(raw_reflectable) {
-                Occupied(entry) => {
-                    if *entry.get().lock().unwrap() != 0 {
-                        // there could have been a new reference taken since
-                        // this message was dispatched.
-                        return;
-                    }
-
-                    let _ = entry.remove();
-                }
-                Vacant(_) => {
-                    // there could be a cleanup message dispatched, then a new
-                    // pinned reference obtained and released before the message
-                    // is processed, at which point there would be no matching
-                    // hashtable entry.
-                    info!("attempt to cleanup an unrecognized reflector");
-                }
-            }
-        })
+/// Remove null entries from the live references table
+fn remove_nulls<K: Eq + Hash + Clone, V> (table: &mut HashMap<K, Weak<V>>) {
+    let to_remove: Vec<K> =
+        table.iter()
+        .filter(|&(_, value)| Weak::upgrade(value).is_none())
```

Comment:
Yep, that's why that API doesn't exist, the synchronization semantics need to be decided by the caller.

As far as racing here goes, that's okay -- a few extra references hanging around as live when they aren't is nbd. It will get caught in the next pass.